File: event-server-ts/src/types/character.ts
import type { Database } from "./db"
import { z } from "zod"

export type Character = Database["public"]["Tables"]["character"]["Insert"]
export type CharacterFull = Database["public"]["Tables"]["character"]["Row"]

export type CharactersEvents = {
  level_ups: CharacterFull[]
  deaths: CharacterFull[]
}

export function normalizeString(str: string): string {
  return str.trim().toLowerCase().replace(/\s+/g, "-")
}

const characterClassEnum = z.preprocess(
  (val) => {
    if (typeof val !== "string") return val
    return normalizeString(val)
  },
  z.enum([
    "death-knight",
    "demon-hunter",
    "druid",
    "evoker",
    "hunter",
    "mage",
    "monk",
    "paladin",
    "priest",
    "rogue",
    "shaman",
    "warlock",
    "warrior",
  ]) satisfies z.ZodType<Database["public"]["Enums"]["character_class"]>,
)

export const gameRegionEnum = z.preprocess(
  (val) => {
    if (typeof val !== "string") return val
    return normalizeString(val)
  },
  z.enum(["us", "eu", "kr", "tw", "cn"]) satisfies z.ZodType<Database["public"]["Enums"]["game_region"]>,
)

export const gameVersionEnum = z.preprocess(
  (val) => {
    if (typeof val !== "string") return val
    return normalizeString(val)
  },
  z.enum(["classic", "tbc-classic", "wrath-classic", "cata-classic", "retail"]) satisfies z.ZodType<
    Database["public"]["Enums"]["game_version"]
  >,
)

export const CharacterSchema = z.object({
  class: characterClassEnum.nullable().optional(),
  level: z.number().optional(),
  zone: z.string().optional(),
  online: z.boolean().optional(),
  died_at: z.number().optional(),
})

export const CharacterRequestSchema = z.object({
  region: gameRegionEnum,
  version: gameVersionEnum,
  realms: z.record(z.record(CharacterSchema)),
})

export const CharacterTelegramSchema = z.object({
  version: gameVersionEnum,
  region: gameRegionEnum,
  realm: z
    .string()
    .min(2, "Realm name must be at least 2 characters")
    .transform((s) => s.charAt(0).toUpperCase() + s.slice(1).toLowerCase()),
  name: z
    .string()
    .min(2, "Character name must be at least 2 characters")
    .transform((s) => s.charAt(0).toUpperCase() + s.slice(1).toLowerCase()),
})

File: event-server-py/app/schemas/character.py
from datetime import datetime
from enum import Enum
from typing import Annotated, Dict, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator


class GameVersion(str, Enum):
    CLASSIC = "classic"
    TBC_CLASSIC = "tbc-classic"
    WRATH_CLASSIC = "wrath-classic"
    CATA_CLASSIC = "cata-classic"
    RETAIL = "retail"


class GameRegion(str, Enum):
    US = "us"
    EU = "eu"
    KR = "kr"
    TW = "tw"
    CN = "cn"


class CharacterClass(str, Enum):
    DEATH_KNIGHT = "death-knight"
    DEMON_HUNTER = "demon-hunter"
    DRUID = "druid"
    EVOKER = "evoker"
    HUNTER = "hunter"
    MAGE = "mage"
    MONK = "monk"
    PALADIN = "paladin"
    PRIEST = "priest"
    ROGUE = "rogue"
    SHAMAN = "shaman"
    WARLOCK = "warlock"
    WARRIOR = "warrior"


class CharacterBase(BaseModel):
    version: GameVersion
    region: GameRegion
    realm: str
    name: str
    class_: Optional[Annotated[CharacterClass, Field(alias="class")]] = None
    level: Optional[int] = None
    zone: Optional[str] = None
    online: Optional[bool] = None
    died_at: Optional[datetime] = None

    model_config = ConfigDict(
        use_attribute_docstrings=True,
        use_enum_values=True,
        populate_by_name=True,
        json_schema_extra={
            "examples": [
                {
                    "realm": "silvermoon",
                    "name": "Thrall",
                    "level": 70,
                    "class_": "warrior",
                    "online": True,
                    "zone": "Orgrimmar",
                    "version": "classic",
                    "region": "us",
                }
            ]
        },
    )


class CharacterCreate(CharacterBase):
    died_at: Optional[int] = None


class Character(CharacterBase):
    id: str

    model_config = ConfigDict(use_attribute_docstrings=True, from_attributes=True)


class CharacterUpdate(BaseModel):
    level: Optional[int] = None
    zone: Optional[str] = None
    online: Optional[bool] = None
    died_at: Optional[int] = None

    model_config = ConfigDict(
        use_attribute_docstrings=True,
        json_schema_extra={"examples": [{"level": 70, "online": False, "zone": "Stormwind"}]},
    )


class CharacterEventData(BaseModel):
    class_: Optional[CharacterClass] = Field(alias="class", default=None)
    level: Optional[int] = None
    zone: Optional[str] = None
    online: Optional[bool] = None
    died_at: Optional[int] = None

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={"example": {"online": True, "level": 23, "class": "warrior", "zone": "Thunder Bluff"}},
    )

    @field_validator("class_", mode="before")
    @classmethod
    def transform_class(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        if isinstance(value, str):
            return value.lower().replace(" ", "-")
        # Handle cases where the value might be the enum itself
        return value


class EventWatcherRequest(BaseModel):
    region: GameRegion
    version: GameVersion
    realms: Dict[str, Dict[str, CharacterEventData]]

    model_config = ConfigDict(
        use_enum_values=True,
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "region": "us",
                "version": "classic",
                "realms": {
                    "Doomhowl": {
                        "Furioustea": {"online": True, "level": 23, "class": "warrior", "zone": "Thunder Bluff"}
                    }
                },
            }
        },
    )
